File: apps/node/src/fizzbuzz/domain/type/fizz-buzz-type.ts
import { FizzBuzzValue } from "../model/fizz-buzz-value.js";
import { FizzBuzzType01 } from "./fizz-buzz-type-01.js";
import { FizzBuzzType02 } from "./fizz-buzz-type-02.js";
import { FizzBuzzType03 } from "./fizz-buzz-type-03.js";
import { FizzBuzzTypeName } from "./fizz-buzz-type-name.js";

export abstract class FizzBuzzType {
  static readonly TYPE_01 = 1;
  static readonly TYPE_02 = 2;
  static readonly TYPE_03 = 3;

  abstract generate(number: number): FizzBuzzValue;

  static create(type: number = FizzBuzzType.TYPE_01): FizzBuzzType {
    switch (type) {
      case FizzBuzzType.TYPE_01:
        return new FizzBuzzType01();
      case FizzBuzzType.TYPE_02:
        return new FizzBuzzType02();
      case FizzBuzzType.TYPE_03:
        return new FizzBuzzType03();
      default:
        throw new Error(`未定義のタイプ: ${type}`);
    }
  }

  static tryCreate(typeName: FizzBuzzTypeName): FizzBuzzType | undefined {
    const typeMap: Record<FizzBuzzTypeName, () => FizzBuzzType> = {
      [FizzBuzzTypeName.TYPE_01]: () => new FizzBuzzType01(),
      [FizzBuzzTypeName.TYPE_02]: () => new FizzBuzzType02(),
      [FizzBuzzTypeName.TYPE_03]: () => new FizzBuzzType03(),
    };

    return typeMap[typeName]?.();
  }
}

File: apps/kotlin/src/main/kotlin/fizzbuzz/domain/type/FizzBuzzType.kt
package fizzbuzz.domain.type

/**
 * FizzBuzz のタイプ（出力の切り替え）を表す列挙型。
 * 各ケースが generate を実装する（State パターン相当）。
 */
enum class FizzBuzzType {
    /** 通常の FizzBuzz。 */
    TYPE_01 {
        override fun generate(n: Int): String = when {
            n % 15 == 0 -> "FizzBuzz"
            n % 3 == 0 -> "Fizz"
            n % 5 == 0 -> "Buzz"
            else -> n.toString()
        }
    },

    /** 数字のみ。 */
    TYPE_02 {
        override fun generate(n: Int): String = n.toString()
    },

    /** Fizz の場合のみ（Buzz なし）。 */
    TYPE_03 {
        override fun generate(n: Int): String = when {
            n % 15 == 0 -> "FizzBuzz"
            n % 3 == 0 -> "Fizz"
            else -> n.toString()
        }
    };

    /** 数 [n] をタイプに従って文字列へ変換する。 */
    abstract fun generate(n: Int): String

    companion object {
        /**
         * タイプ番号から FizzBuzzType を生成する。未定義の番号は失敗を返す。
         */
        fun create(no: Int): Result<FizzBuzzType> = when (no) {
            1 -> Result.success(TYPE_01)
            2 -> Result.success(TYPE_02)
            3 -> Result.success(TYPE_03)
            else -> Result.failure(IllegalArgumentException("該当するタイプは存在しません: $no"))
        }
    }
}
